### ISSUE TITLE
Doesn't work inside a virtualenv

### ISSUE THREAD

**Issue by raxod502** [jorgenschaefer/elpy#1546]
# Summary

Autocompletion works as desired when I _don't_ activate a virtualenv, as I've installed `jedi` globally. However, as soon as I _do_ activate virtualenv, autocompletion suddenly stops working at all, as I do not have `jedi` installed in my virtualenv (it's for a shared project managed via Pipenv).

I can see why this happens—if Elpy tries to run `jedi` via whatever `python` is available on the PATH (which would be the one from my virtualenv) then of course it won't be able to find `jedi`. But this behavior seems like a bug to me. Why isn't Elpy running `jedi` through the global `python`?

I don't know how to work around this except by installing all of my project dependencies globally or by installing all of Elpy's dependencies into my project's `Pipfile`, both of which are totally horrifying options.

It seems like #1020 is reporting the same problem but no solution is presented (unless there is some way to automatically install extra packages into every virtualenv without interfering with anything?).

# Steps to reproduce

* `M-x pyvenv-workon`
* Try to use autocompletion

# My configuration
## OS

macOS 10.14.3

## Result of `(elpy-config)`

_Before_ activating virtualenv:

```
Virtualenv........: backend-hrC0GD4S (/Users/raxod502/.virtualenvs/backend-hrC0GD4S)
RPC Python........: 3.7.2 (/usr/local/bin/python3)
Interactive Python: python3 (/usr/local/bin/python3)
Emacs.............: 27.0.50
Elpy..............: 1.26.0
Jedi..............: 0.13.3
Rope..............: 0.12.0
Autopep8..........: 1.4.3
Yapf..............: 0.26.0
Black.............: 18.9b0
Syntax checker....: flake8 (/Users/raxod502/.local/bin/flake8)
```

_After_ activating virtualenv:

```
Virtualenv........: backend-hrC0GD4S (/Users/raxod502/.virtualenvs/backend-hrC0GD4S)
RPC Python........: 3.7.2 (/Users/raxod502/.virtualenvs/backend-hrC0GD4S/bin/python3)
Interactive Python: python3 (/Users/raxod502/.virtualenvs/backend-hrC0GD4S/bin/python3)
Emacs.............: 27.0.50
Elpy..............: 1.26.0
Jedi..............: Not found (0.13.3 available)
Rope..............: Not found (0.12.0 available)
Autopep8..........: Not found (1.4.3 available)
Yapf..............: Not found (0.26.0 available)
Black.............: Not found (18.9b0 available)
Syntax checker....: flake8 (/Users/raxod502/.local/bin/flake8)
```

## Elpy configuration in my init.el

```
(setq elpy-modules (remq 'elpy-module-highlight-indentation elpy-modules))
(setq elpy-modules (remq 'elpy-module-flymake elpy-modules))
(setq elpy-rpc-python-command python-shell-interpreter)
(elpy-enable)
```

**Comment by gopar** [comment]
Hey, I use `pipenv` at work too and the way I get around it is by simply running `pip install jedi`. Pipenv is just a wrapper around pip so pip should still be available for you to use, just don't include it in the `Pipefile` :P

**Comment by raxod502** [comment]
You're right—that works. But I do have to do that for every single virtualenv that I ever use, and I have to redo it if I ever use Pipenv to manage my dependencies. It's a plausible _temporary workaround_, but an upstream fix is definitely necessary IMO :)

FWIW, anaconda-mode with pipenv-mode doesn't have this problem (according to a friend). I haven't yet figured out how anaconda-mode is invoking Jedi so as to avoid it.

**Comment by galaunay** [comment]
That is actually a good point.

Jedi supports using different virtual environment so it should indeed be started from the global python.
I started a PR (#1548) to change that.

It is apparently working but need to be thoroughly tested.

**Comment by gopar** [comment]
Ohh didn't know jedi supported that. Personally, I made editted the `postmkvirtualenv` or whatever file is called to install elpy deps and other tools whenever I create a new environment

**Comment by galaunay** [comment]
We will have to decide if it worth the shot.
It could be painful to change how the rpc works (a lot of potential for new bugs I am afraid).

**Comment by avocadoras** [comment]
Hi. I struggled figuring out a solution to this but there is no easy way to do this. The issue is also mentioned here https://github.com/jorgenschaefer/elpy/wiki/FAQ#q-do-i-have-to-install-all-these-packages-every-time. 

I found myself installing `pyenv-virtualenvwrapper` as well since it seems I cannot get `togglesitepackages` to work well with `virtualenvwrapper`.

Also, is there a way to add some features like those in `virtualenvwrapper.el`? I like the idea of not having to `pip install virtualenvwrapper` and working straight from Emacs to create and manage environments. Also, I have issues working with flycheck in virtual environments and cannot get it to work. I had to install `pyenv-mode` to solve the problem.

**Comment by PeterChristoph** [comment]
Hello,
I have got the same problem, autocompletion works fine before activating virtualenv, and not working after activating virtualenv. Also, after activating virtualenv, Elpy does not find Jedi. 

The difference is that I use a conda environment and that I installed Jedi in the environment:

Elpy-config before activating virtualenv:

Virtualenv........: None
RPC Python........: 3.6.4 (/opt/anaconda/bin/ipython)
Interactive Python: ipython (/opt/anaconda/bin/ipython)
Emacs.............: 25.2.2
Elpy..............: 1.28.0
Jedi..............: 0.13.3
Rope..............: 0.10.7
Autopep8..........: 1.4.4
Yapf..............: 0.27.0
Black.............: Not found (19.3b0 available)
Syntax checker....: flake8 (/opt/anaconda/bin/flake8)


Elpy-config after activating virtualenv:

Virtualenv........: dyps (/opt/anaconda3/envs/dyps)
RPC Python........: 3.6.2 (/opt/anaconda3/envs/dyps/bin/ipython)
Interactive Python: ipython (/opt/anaconda3/envs/dyps/bin/ipython)
Emacs.............: 25.2.2
Elpy..............: 1.28.0
Jedi..............: Not found (0.13.3 available)
Rope..............: 0.9.4-1
Autopep8..........: 1.4.4
Yapf..............: 0.27.0
Black.............: 19.3b0
Syntax checker....: flake8 (/opt/anaconda3/envs/dyps/bin/flake8)

**Comment by galaunay** [comment]
That seems to be a different problem.

Elpy checks if jedi is installed just by trying to import it.
From inside emacs and with the virtualenv activated, can you import jedi ? by just evaluating a python buffer containing `import jedi` for example.

**Comment by PeterChristoph** [comment]
Thank you very much. That is the problem. I am getting an "ImportError: cannot import name 'splitlines'".

I reinstalled Jedi into the virtual environment and Elpy-config now finds Jedi. The weird thing is that Company still does not find any completion when in virtual env.

my Elpy-config after acitvating virtualenv:
Elpy Configuration

Virtualenv........: dyps (/opt/anaconda3/envs/dyps)
RPC Python........: 3.6.2 (/opt/anaconda3/envs/dyps/bin/python)
Interactive Python: ipython (/opt/anaconda3/envs/dyps/bin/ipython)
Emacs.............: 25.2.2
Elpy..............: 1.28.0
Jedi..............: 0.14.0
Rope..............: 0.9.4-1
Autopep8..........: 1.4.4
Yapf..............: 0.27.0
Black.............: 19.3b0
Syntax checker....: flake8 (/opt/anaconda3/envs/dyps/bin/flake8)

**Comment by galaunay** [comment]
There is an recent [issue](https://github.com/davidhalter/jedi/issues/1348) with jedi 0.14.
Would it fix your issue if you install parso==0.5 in the virtualenv ?

**Comment by PeterChristoph** [comment]
parso==0.5 is already installed in virtualenv. I also tried an older version. I think it was 0.10.... 
Anyway, thank you for your great work!
